students/idcrickmore/lesson05/mailroom_pt3.py

```diff
@@ -0,0 +1,140 @@
+type 'menu' at any time to return to the menu""")
+
+    # asks for a name from the user, or for the 'list' prompt
+    name_check = input("Enter first and last name of a donor\n\n-> ").lower()
+
+    try:
+        options.get(name_check)()
+    except TypeError:
+        try:
+            donation = float(input("What is the donation amount? -> "))
+        except ValueError:
+            print("\nMUST ENTER A NUMBER")
+            thankyou()
+        else:
+            name_check = name_check.title()
+            # check to see if 'name_check' is in the donor dict...
+            # if it is, append it
+            try:
+                donors[name_check].append(donation)
+            except KeyError:
+                # if 'name_check' isn't in the donor dict...
+                # set  the default format value as a list ([]) and append it
+                donors.setdefault((name_check), []).append(donation)
+            finally:
+                letter_dict = {"name": name_check, "don": donation}
+                print("""\n------------------ Letter ------------------\n
+Dear {name},
+
+You rock. Your fat contribution of ${don:,.2f}
+will go a long way to lining my pockets.
+
+Sincerely,
+Scrooge McDuck""".format(**letter_dict))
+
+                menu()
+
+
+def letters():
+    print("""\n--------- Exporting Letters to Everyone ---------\n
+Thank you letters have been exported to {}""".format(os.getcwd()))
+
+    for name in donors:
+        letter_dict = {"donor": name, "donation": donors[name][-1]}
+        letter = """Dear {donor},
+
+Your most recent contribution of ${donation:,.2f} to my money vault
+is very appreciated. I will spend it freely but wisely.
+
+Sincerely,
+Scrooge McDuck""".format(**letter_dict)
+        with open(str(name) + "_" + str(datetime.date.today()) +
+                  '.txt', 'w') as out_file:
+            out_file.write(letter)
+
+    menu()
+
+
+def donor_list():
+    # lists all the donors
+    # calls the 'thankyou' function at the end to ask for user input
+    print("\n------------------ DONOR LIST ------------------\n")
+
+    for name in donors:
+        print(name)
+
+    thankyou()
+
+
+def report():
+    print("\n-------------------- REPORT --------------------\n")
+
+    column = ["Donor Name", "| Total Given", "| Num Gifts", "| Average Gift"]
+    donors_report = [[name, sum(donors[name]), len(donors[name]),
+                     (sum(donors[name])/len(donors[name]))]
+                     for name in donors]
+
+    # and reverses the order from largest to smallest
+    donors_report = sorted(donors_report, key=itemgetter(1), reverse=True)
+
+    print("{:<15}{:>17}{:>15}{:>10}".format(*column))
+    print("---------------------------------------------------------------")
+
+    # loops through 'donors_report' and
+    # dumps all values for 'name' into .format
+    for name in donors_report:
+        print('{:<20} ${:>13.2f}{:>12}  ${:>10.2f}'.format(*name))
+
+    menu()
```

Comment:
from extendability point of view, it makes more sense to have while True loop with redirection to "work" functions.

Imagine you need to add another functionality option, whoever adds that function needs to remember to call `main` on it which is error prone.

If you had while True loop, whoever would extend the program would only need to consider new functionality and not worry about the main program flow.